nv2a: Adjust NaN handling for fog coordinate and colors to be similar to HW

abaire opened the pull request:
Resurrecting #913 and incorporating some of the comments from #1780 in the hopes that:

1) This approach will work across different GPU manufacturers (testing needed, it works for M3 mac and my old NVIDIA 1070)
2) That the handful of fog nan mishandles that have been identified are the only places where NaN handling is important in practice so the perf hit of patching up NaNs can be limited to once per vertex instead of having to run on all attribs.

I suspect that the MUL zero handling fix that was added in #1045 may already be preventing/masking some issues.

[HW exceptional fog tests](https://abaire.github.io/nxdk_pgraph_tests_golden_results/results/Fog_exceptional_value/index.html)
[PR test results](https://abaire.github.io/xemu-dev_pgraph_test_results/fix_nan_fog/index.html)

This PR does not fix all issues with the exceptional fog tests, but it does fix the ones that seem to matter for Otogi. I have not fully reversed how fog params are utilized so I think the pgraph tests may be generating cases that would never be triggered by D3D.


(Probably) Fixes https://github.com/xemu-project/xemu/issues/365 by forcing treatment of NaN fog values generated in the vertex shader to follow the same pattern as HW.
Reportedly fixes #1026 as well

Triticum0 commented:
Thank abaire. You been working really hard with the pr's keep up the good work.

viniciusol263 commented:
Otogi 1 working perfectly here on my rig: RTX 2070 SUPER

Enkak commented:
Tried 3 games from the same company, From Software, that had the issue of no textures. All 3 working good/normal here with an old GPU NVIDIA GTX 765M regarding that issue with this PR:
Otogi 1
Otogi 2
Thousand Land

Fabxx commented:
Tested on splinter cell double agent and pandora tomorrow, no regressions to report.

real-guilty commented:
I've been trying to build this in MSYS2 for the last couple hours, but even after pulling _all_ the needed dependencies, CMake and Vulkan has done everything it can to prevent me from testing it for myself. They should have split the Vulkan backend into its own version of xemu--implementing it this way has been nothing but problems. For the third or fourth time now, I'm giving up on this. Hopefully your pull is distributable at some point.

real-guilty commented:
> Any reason the build artifacts attached to the PR don't work for you? There are windows builds.

I am Git-illiterate. I found the artifacts though and am running O1 now. Looks great--0.8.47 and I can see everything. Will run some stages--hell, maybe speedrun it--then go through O2 just to be another name that can confirm the fix.

You wouldn't happen to also be wise on the audio utilization issues I've reported, would you? "Clouded Moon" and "Inner Sanctum" for O1, and "Golden Sea" for O2.

mborgerson commented:
@real-guilty 

> I've been trying to build this in MSYS2 for the last couple hours, but even after pulling all the needed dependencies, CMake and Vulkan has done everything it can to prevent me from testing it for myself. They should have split the Vulkan backend into its own version of xemu--implementing it this way has been nothing but problems. For the third or fourth time now, I'm giving up on this. Hopefully this is merged or you release this yourself at some point.

Users are recommend to use the release build or download artifacts from GitHub instead of building from source. I'm not sure what issues you're having with building from source, but it sounds like you didn't follow [the directions](https://xemu.app/docs/dev/building-from-source/).

> You wouldn't happen to also be wise on the audio utilization issues I've reported, would you? "Clouded Moon" and "Inner Sanctum" for O1, and "Golden Sea" for O2.

Let's keep the conversation on this issue focused on the matter at hand and various other Otogi issues. General chat is welcome on the xemu Discord server.

abaire commented:
Thanks to `Sinful` on Discord: the current fix does not work on SteamDeck in Windows

```
        "{\n"
        "  bvec4 nans = isnan(src);\n"
        "  if (!any(nans)) {\n"
        "    return src;\n"
        "  }\n"
        "  return mix(src, sign(floatBitsToInt(src)), nans);\n"
        "}
```

abaire commented:
With `Sinful`'s help I was able to get the fix working on the Steam Deck. Will need to be re-tested on NVIDIA HW as well to make sure it hasn't regressed.

At the moment I can confirm that it works on Darwin/M3 (but the default NaN handling on Darwin/M3 already worked without this PR).

Triticum0 commented:
I'll test it. post results in 20 mins

viniciusol263 commented:
![image](https://github.com/user-attachments/assets/7a5010a0-2b27-4d0e-ba1b-9631b3ac7de5)
on RTX 2070 SUPER

Triticum0 commented:
You still need to have a look at Pro Cast: Sports Fishing Game  it still doesn't look correct

Triticum0 commented:
Here you go. GPU:3060 ti

Otogi 1
![xemu-2025-04-15-23-26-36](https://github.com/user-attachments/assets/eca1b668-93a8-407c-a767-bb2b66501c00)

Otogi 2
![xemu-2025-04-15-23-27-34](https://github.com/user-attachments/assets/5975f486-52b0-4167-baa7-1dc49e15d5bb)

Thousand Land
![xemu-2025-04-15-23-29-32](https://github.com/user-attachments/assets/fc8d2583-73b2-4a25-a281-f6a61d01db6a)

Triggerman
![xemu-2025-04-15-23-30-54](https://github.com/user-attachments/assets/ed26eb83-d122-4058-8cc8-a4ddae07e662)


Pro Cast: Sports Fishing
![xemu-2025-04-15-23-26-23](https://github.com/user-attachments/assets/109400a8-d78b-4c90-aa38-eb6158f3e3f6)

abaire commented:
> You still need to have a look at Pro Cast: Sports Fishing Game it still doesn't look correct

@Triticum0 I'm assuming the issue there has some cause other than NaN fog. I've ordered the game and will investigate in the coming weeks. If there isn't already a GH issue specifically tracking the problem, can you create one?

Triticum0 commented:
Not sure if that necessary coldhex has the game and posted about on previous pr https://github.com/xemu-project/xemu/pull/1780#issuecomment-2480620232

Triticum0 commented:
If that not helpful then i happy make an new issue tomorrow.

abaire commented:
Unfortunately it seems that the fix that worked on my personal branch did not work after being re-applied here, working w/ `Sinful` to see if I can figure out why. Back to draft until then.

@Triticum0 let's add a new issue specific for Pro Cast, it'll require applying this fix to the color channels as well as fog

abaire commented:
Turns out that the `floatBitsToInt` causes this to fail on AMD (at least on Steam Deck). Removed and added a comment for future investigation. We'll want to test this yet again on NVIDIA HW.

viniciusol263 commented:
Working on a RTX 2070 SUPER 
![image](https://github.com/user-attachments/assets/1c949df9-cc6d-4fc3-ac21-a3dc1e4bb9ed)

abaire commented:
Something changed with macOS 15.4.1 such that `sign(nan)` appears to be returning 0 now. I think we can drop the sign check and just set all nan's to 1. It won't be exactly like the HW results since multiplying nan's by 0/practically 0 numbers returns 0, but it will match most of the cases, including Otogi.

abaire commented:
Something changed with macOS 15.4.1 such that `sign(nan)` appears to be returning 0 now. I think we can drop the sign check and just set all nan's to 1. It won't be exactly like the HW results since multiplying nan's by 0/practically 0 numbers returns 0, but it will match most of the cases, including Otogi.

abaire commented:
Verified working on Darwin, NVIDIA, and AMD (Steam Deck)

Triticum0 commented:
Glad it works now on all vendors. hopefully merged soon.

mborgerson commented on hw/xbox/nv2a/pgraph/glsl/vsh.c:
```diff
@@ -128,6 +128,21 @@ MString *pgraph_gen_vsh_glsl(const ShaderState *state, bool prefix_outputs)
+        "vec4 _in(vec4 v) { return v.xyzw; }\n"
+        "\n"
+        "#define FixNaNClamped(src, mask) _FixNaNClamped(_in(src)).mask\n"
+        "vec4 _FixNaNClamped(vec4 src)\n"
```
What values are clamped here, to what bounds? A better name might be `NanToOne`, or something

abaire replied:
Sounds good, the name hasn't kept up with the evolution to work around implementation-specific issues.

abaire commented:
Updated to resolve comments and expanded to cover color channels as discussed. 

This introduces some new/modified deltas versus hardware in lighting test results on Darwin/M3, but only in exceptional cases (e.g., infinite attenuation and spotlight falloff due to pathological values that I think cannot be generated by DirectX). The dev branch results page has been updated to demonstrate.

Enkak commented:
Hi and just a quick question if this has been merged to master yet? 

I have the github artifact just for the Otogi games but since there were some updates to the PR just to check if maybe I missed the merging or if there's still some technical conflicts or issues to solve.

Cheers and thank you all for the hard work.

abaire commented:
> Hi and just a quick question if this has been merged to master yet?

When it's merged this page will have a purple icon that says "Merged" instead of a green icon that says "Open".

Enkak commented:
> > Hi and just a quick question if this has been merged to master yet?
> 
> When it's merged this page will have a purple icon that says "Merged" instead of a green icon that says "Open".

Ah I see. Many thanks for the clarification!

Triticum0 commented:
Tested Dinosaur Hunter with pr and does seem to have any affect on issue #1622

antoniodesousa commented:
I tested all the games that are known to have the black fog issue and they're all fixed in this PR. I also tested a bunch of other games and I didn't see any apparent regressions. Hopefully, the PR gets merged soon. I'm tired of using a custom build to play the Otogi games.

```
Version: 0.0.0
Branch: fix_nan_fog
Commit: d6b00e952f01a340be61081947cb326e4666c707
Date: Sat Apr 26 23:08:27 UTC 2025
CPU: AMD Ryzen 9 6900HX with Radeon Graphics        
OS Platform: Windows
OS Version: 24H2
Manufacturer: ATI Technologies Inc.
GPU Model: AMD Radeon RX 6600M
Driver: 4.0.0 Core Profile Context 25.5.1.250417
Shader: 4.60
```
![xemu-2025-05-15-15-42-00](https://github.com/user-attachments/assets/5ff2da86-4127-4573-b948-a651fa2f5304)
![xemu-2025-05-15-15-47-15](https://github.com/user-attachments/assets/35860551-11b2-4a32-9b46-bddd9f8dbda2)
![xemu-2025-05-15-15-47-49](https://github.com/user-attachments/assets/f8310a47-fa91-4962-b559-90526ff1d4a2)
![xemu-2025-05-15-15-49-58](https://github.com/user-attachments/assets/707dea0f-0aac-492b-8e11-f49272442fe2)
![xemu-2025-05-15-16-07-58](https://github.com/user-attachments/assets/35f3288a-a4cb-4e30-9b46-66b55b76e539)

mborgerson commented on hw/xbox/nv2a/pgraph/glsl/vsh.c:
```diff
@@ -128,6 +128,14 @@ MString *pgraph_gen_vsh_glsl(const ShaderState *state, bool prefix_outputs)
+
+    mstring_append(
+        header,
+        "vec4 NaNToOne(vec4 src)\n"
```
nit: Let's just move this up to follow `clampAwayZeroInf` definition above

abaire replied:
Done

abaire commented on hw/xbox/nv2a/pgraph/glsl/vsh.c:
```diff
@@ -232,17 +240,17 @@ MString *pgraph_gen_vsh_glsl(const ShaderState *state, bool prefix_outputs)
         }
 
-        mstring_append(body, "  oFog.xyzw = vec4(fogFactor);\n");
+        mstring_append(body, "  oFog.xyzw = NaNToOne(vec4(fogFactor));\n");
```
Done

mborgerson commented:
Thanks! Adjusted commit message to be more descriptive of the change

cerroverb commented:
no fucking way this was actually finally fixed
